File: back/src/Game.ts
import config from './config';
import Snake from './entities/Snake';
import { Server, Socket } from 'socket.io';
import axios from 'axios';
import Food from './entities/Food';
import GameEntity from './entities/GameEntity';
import * as shortid from 'shortid';

const schemapack = require('schemapack');
const simplify = require('simplify-js');

const gameUpdate = schemapack.build({
  snakes: [
    {
      id: 'string',
      x: 'int16',
      y: 'int16',
      isBlinking: 'bool',
      isSpeedUp: 'bool',
      length: 'uint16',
      scale: 'float32',
      fillColor: 'string',
      username: 'string',
      angle: 'float32',
      points: [
        {
          x: 'int16',
          y: 'int16',
        },
      ],
      width: 'float32',
    },
  ],
  foods: [
    {
      id: 'string',
      x: 'int16',
      y: 'int16',
      width: 'float32',
      height: 'float32',
    },
  ],
});

const {
  MAP_HEIGHT,
  MAP_RECT_HEIGHT,
  MAP_RECT_WIDTH,
  MAP_WIDTH,
  SOCKET_PORT,
  GAMELOOP_RATE,
  INITIAL_FOOD_COUNT,
  INITIAL_FOOD_VALUE,
  ORCHESTRATOR_URL,
  ORCHESTRATOR_SECRET,
} = config;

const collision = (
  dom: {
    x: number;
    y: number;
    width?: number;
    height?: number;
  },
  dom2: {
    x: number;
    y: number;
    width?: number;
    height?: number;
  },
  isRect?: boolean,
): boolean => {
  const disX = dom.x - dom2.x;
  const disY = dom.y - dom2.y;
  const dw = dom.width + dom2.width;

  if (Math.abs(disX) > dw || Math.abs(disY) > dom.height + dom2.height) {
    return false;
  }

  return isRect ? true : Math.hypot(disX, disY) < dw / 2;
};

// Map class
export class Game {
  public snakes: Snake[];
  public count: number;
  public foods: Food[];

  constructor(
    private wss: Server,
    public width: number = MAP_WIDTH,
    public height = MAP_HEIGHT,
  ) {
    this.snakes = [];
    this.count = 0;
    this.foods = [];
    for (let i = 0; i < INITIAL_FOOD_COUNT; i += 1) {
      this.foods.push(
        new Food({
          id: shortid.generate(),
          x: ~~(Math.random() * (MAP_WIDTH - 100) + 100 / 2),
          y: ~~(Math.random() * (MAP_HEIGHT - 100) + 100 / 2),
          size: 10,
          value: INITIAL_FOOD_VALUE,
        }),
      );
    }
  }

  private confirmConnection() {
    axios.get(
      `http://${ORCHESTRATOR_URL}/confirm/${process.env.EB_SERVER_ID}`,
      {
        auth: {
          username: '',
          password: ORCHESTRATOR_SECRET,
        },
        withCredentials: true,
      },
    );
  }

  private disconnect() {
    axios.get(
      `http://${ORCHESTRATOR_URL}/disconnect/${process.env.EB_SERVER_ID}`,
    );
  }

  public init() {
    this.wss.listen(SOCKET_PORT);
    setInterval(() => this.run(), GAMELOOP_RATE);

    this.wss.on('connection', (ws: Socket) => {
      this.confirmConnection();

      let currentPlayer: Snake;

      ws.on('register', (username: string) => {
        currentPlayer = new Snake({
          username,
          id: shortid.generate(),
          size: 30,
          length: 280,
          angle: Math.random() * 2 * Math.PI,
          x: ~~(Math.random() * (MAP_WIDTH - 100) + 100 / 2),
          y: ~~(Math.random() * (MAP_HEIGHT - 100) + 100 / 2),
        });
        this.snakes.push(currentPlayer);
        ws.emit('register-success', currentPlayer);
      });

      ws.on('move', (orientation: any) => {
        if (currentPlayer) currentPlayer.moveTo(orientation.x, orientation.y);
      });

      ws.on('player-speed-up', () => {
        currentPlayer.speedUp();
      });

      ws.on('player-speed-down', () => {
        currentPlayer.speedDown();
      });

      ws.on('disconnect', () => {
        if (currentPlayer) {
          const idx = this.snakes.findIndex(
            snake => snake.id === currentPlayer.id,
          );
          if (idx !== -1) {
            // handle the ctrl+R refresh case
            this.snakes.splice(idx, 1);
            currentPlayer = undefined;
          }
        }
        this.disconnect();
      });
    });
  }

  private run() {
    const snakeDeleted: Map<string, Snake> = new Map();

    this.snakes.forEach((snake: Snake) => {
      snake.update();

      if (snake.length < 0) {
        snakeDeleted.set(snake.id, snake);
        return;
      }

      this.snakes.forEach((snake2: Snake) => {
        snake2.points.forEach((point: { x: number; y: number }) => {
          if (
            snake2.id !== snake.id &&
            collision(snake, {
              ...point,
              width: snake2.width,
              height: snake2.height,
            })
          ) {
            if (!snakeDeleted.has(snake.id)) {
              snakeDeleted.set(snake.id, snake);
            }
          }
        });
      });

      this.foods.forEach((food: Food) => {
        food.update();
        if (!collision(snake, food)) {
          return;
        }
        const added = snake.eat(food);
        this.foods.splice(this.foods.indexOf(food), 1);

        const newScale = snake.scale + added / (snake.width * 4);
        if (newScale < 1.4) {
          snake.scale = newScale;
        }
      });

      const diff = INITIAL_FOOD_COUNT / 2 - this.foods.length;

      for (let i = 0; i < diff; i += 1) {
        this.foods.push(
          new Food({
            id: shortid.generate(),
            x: ~~(Math.random() * (MAP_WIDTH - 100) + 100 / 2),
            y: ~~(Math.random() * (MAP_HEIGHT - 100) + 100 / 2),
            size: 10,
            value: INITIAL_FOOD_VALUE,
          }),
        );
      }

      this.limit(snake);
    });

    snakeDeleted.forEach((snake: Snake) => {
      this.snakes.splice(this.snakes.indexOf(snake), 1);
      snake.points.forEach(
        (point, index) =>
          index % 10 === 0 &&
          this.foods.push(
            new Food({
              id: shortid.generate(),
              x: point.x,
              y: point.y,
              size: 10,
              value: INITIAL_FOOD_VALUE,
            }),
          ),
      );
    });

    this.wss.emit(
      'game-update',
      gameUpdate.encode({
        // snakes: this.snakes.map((snake: Snake) => {
        //   return {
        //     ...snake,
        //     points:
        //       snake.points.length < 20
        //         ? snake.points
        //         : snake.points.filter((point, index) => index % 10 === 0),
        //   };
        // }),
        snakes: this.snakes.map((snake: Snake) => {
          return {
            ...snake,
            points: simplify(snake.points, 5),
          };
        }),
        foods: this.foods,
      }),
    );
  }

  // limit element, prevent it moving to outside
  public limit(element: {
    x: number;
    y: number;
    width?: number;
    height?: number;
  }): void {
    const whalf: number = (element.width || 1) / 2;
    if (element.x < whalf) {
      element.x = whalf;
    } else if (element.x + whalf > this.width) {
      element.x = this.width - whalf;
    }

    const hhalf: number = (element.height || 1) / 2;
    if (element.y < hhalf) {
      element.y = hhalf;
    } else if (element.y + hhalf > this.height) {
      element.y = this.height - hhalf;
    }
  }
}

File: back-go/game.go
package main

import (
	"log"
	"math"
	"net/http"
	"os"
	"time"

	"github.com/gorilla/websocket"
)

type Game struct {
	Count       int
	Snakes      []*Snake
	Foods       []*Food
	Width       float64
	Height      float64
	Connections map[*websocket.Conn]string
}

func NewGame() *Game {
	game := &Game{
		Count:       0,
		Snakes:      []*Snake{},
		Foods:       []*Food{},
		Width:       MAP_WIDTH,
		Height:      MAP_HEIGHT,
		Connections: map[*websocket.Conn]string{},
	}

	for i := 0; i < INITIAL_FOOD_COUNT; i++ {
		game.Foods = append(game.Foods, NewRandomFood())
	}

	return game
}

func collision(ent1, ent2 *GameEntity, isRect bool) bool {
	disX := ent1.X - ent2.X
	disY := ent1.Y - ent2.Y
	dw := ent1.Width + ent2.Width

	if math.Abs(disX) > dw || math.Abs(disY) > ent1.Height+ent2.Height {
		return false
	}

	if isRect {
		return true
	}

	return math.Hypot(disX, disY) < dw/2
}

func rectCollision(rect1, rect2 *Rect) bool {
	return !(rect2.MinX > rect1.MaxX ||
		rect2.MaxX < rect1.MinX ||
		rect2.MaxY < rect1.MinY ||
		rect2.MinY > rect1.MaxY)
}

func checkCollision(s1, s2 *Snake) bool {
	for _, point := range s2.Points {
		if collision(&s1.GameEntity,
			&GameEntity{
				Point:  Point{point.X, point.Y},
				Width:  s2.Width,
				Height: s2.Height,
			},
			false) {
			return true
		}
	}

	return false
}

func confirmConnection() {
	client := &http.Client{}
	req, _ := http.NewRequest("GET", "http://"+ORCHESTRATOR_URL+"/confirm/"+os.Getenv("EB_SERVER_ID"), nil)
	req.SetBasicAuth("", ORCHESTRATOR_SECRET)
	client.Do(req)
}

func disconnect() {
	client := &http.Client{}
	req, _ := http.NewRequest("GET", "http://"+ORCHESTRATOR_URL+"/disconnect/"+os.Getenv("EB_SERVER_ID"), nil)
	client.Do(req)
}

var upgrader = websocket.Upgrader{
	CheckOrigin: func(r *http.Request) bool {
		return true
	},
}

func (g *Game) Init() {
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		log.Println("request")

		conn, err := upgrader.Upgrade(w, r, nil)
		if err != nil {
			log.Println("request failed", err)
			return
		}

		log.Println("connect")

		confirmConnection()

		var currentPlayer *Snake

		for {
			messageType, p, err := conn.ReadMessage()
			if err != nil {
				log.Println("disconnect")
				if currentPlayer != nil {
					idx := FindSnakeIndex(g.Snakes, currentPlayer.Id)
					if idx != -1 {
						g.Snakes = append(g.Snakes[:idx], g.Snakes[idx+1:]...)
					}
				}
				delete(g.Connections, conn)
				conn.Close()
				disconnect()
				return
			}

			var incMsg IncMsg
			_, err = incMsg.Unmarshal(p)
			if err != nil {
				log.Println("error while decoding", err)
				return
			}

			switch incMsg.Topic {
			case "register":
				currentPlayer = NewSnake(incMsg.Username)
				g.Connections[conn] = currentPlayer.Id
				g.Snakes = append(g.Snakes, currentPlayer)

				gameUpdateMsg := &GameUpdateMsg{
					Topic: "register-success",
					Foods: []*FoodMsg{},
					Snakes: []*SnakeMsg{&SnakeMsg{
						Id:            currentPlayer.Id,
						X:             currentPlayer.X,
						Y:             currentPlayer.Y,
						Width:         currentPlayer.Width,
						IsSpeedUp:     currentPlayer.IsSpeedUp,
						FillColor:     currentPlayer.FillColor,
						Angle:         currentPlayer.Angle,
						Scale:         currentPlayer.Scale,
						IsBlinking:    currentPlayer.IsBlinking,
						CollisionRect: currentPlayer.CollisionRect,
						Username:      currentPlayer.Username,
						Speed:         currentPlayer.Speed,
						Length:        currentPlayer.Length,
						Score:         int64(len(currentPlayer.Points)),
						Points:        currentPlayer.SimplifiedPoints,
					}},
				}

				msg := make([]byte, 15000)
				msgLen := gameUpdateMsg.MarshalTo(msg)

				err = conn.WriteMessage(messageType, msg[:msgLen])
				if err != nil {
					log.Println("error while writing message", err)
					return
				}
			case "move":
				if currentPlayer != nil {
					currentPlayer.MoveTo(incMsg.X, incMsg.Y)
				}
			case "player-speed-up":
				currentPlayer.SpeedUp()
			case "player-speed-down":
				currentPlayer.SpeedDown()
			default:
				log.Println("error while reading topic")
			}
		}
	})

	ticker := time.NewTicker(16 * time.Millisecond)
	go func() {
		for _ = range ticker.C {
			g.Run()
		}
	}()

	port := os.Getenv("EB_SERVER_PORT")
	if port == "" {
		port = "4242"
	}

	log.Println("Serving at localhost:" + port)

	http.ListenAndServe(":"+port, nil)
}

func (g *Game) limit(snake *Snake) {
	whalf := snake.Width
	if whalf == 0 {
		whalf = 1
	}
	whalf /= 2

	if snake.X < whalf {
		snake.X = whalf
	} else if snake.X+whalf > g.Width {
		snake.X = g.Width - whalf
	}

	hhalf := snake.Height
	if hhalf == 0 {
		hhalf = 1
	}
	hhalf /= 2

	if snake.Y < hhalf {
		snake.Y = hhalf
	} else if snake.Y+hhalf > g.Height {
		snake.Y = g.Height - hhalf
	}
}

func FindFoodIndex(foods []*Food, id string) int {
	for i, food := range foods {
		if id == food.Id {
			return i
		}
	}
	return -1
}

func FindSnakeIndex(snakes []*Snake, id string) int {
	for i, snake := range snakes {
		if id == snake.Id {
			return i
		}
	}
	return -1
}

func (g *Game) Run() {
	deletedSnakes := map[string]*Snake{}

	for _, snake := range g.Snakes {
		snake.Action()

		if snake.Length < 0 {
			deletedSnakes[snake.Id] = snake
			continue
		}

		for _, snake2 := range g.Snakes {
			if snake2.Id != snake.Id &&
				rectCollision(snake.CollisionRect, snake2.CollisionRect) &&
				checkCollision(snake, snake2) &&
				deletedSnakes[snake.Id] == nil {
				deletedSnakes[snake.Id] = snake
			}
		}

		for _, food := range g.Foods {
			food.Action()

			if !collision(&snake.GameEntity, &food.GameEntity, false) {
				continue
			}

			added := snake.Eat(food)
			idx := FindFoodIndex(g.Foods, food.Id)

			if idx == -1 {
				continue
			}

			g.Foods = append(g.Foods[:idx], g.Foods[idx+1:]...)

			newScale := snake.Scale + added/(snake.Width*4)
			if newScale < 1.4 {
				snake.Scale = newScale
			}
		}

		diff := INITIAL_FOOD_COUNT/2 - len(g.Foods)
		for i := 0; i < diff; i++ {
			g.Foods = append(g.Foods, NewRandomFood())
		}

		g.limit(snake)
	}

	for _, snake := range deletedSnakes {
		idx := FindSnakeIndex(g.Snakes, snake.Id)
		g.Snakes = append(g.Snakes[:idx], g.Snakes[idx+1:]...)
		for i, point := range snake.Points {
			if i%10 == 0 {
				g.Foods = append(g.Foods, NewFood(point.X, point.Y))
			}
		}
	}

	for conn, id := range g.Connections {
		idx := FindSnakeIndex(g.Snakes, id)

		gameUpdateMsg := &GameUpdateMsg{
			Topic:  "game-update",
			Foods:  []*FoodMsg{},
			Snakes: []*SnakeMsg{},
		}

		if idx != -1 {
			snake := g.Snakes[idx]
			for _, food := range g.Foods {
				if rectCollision(snake.DisplayRect, food.CollisionRect) {
					gameUpdateMsg.Foods = append(gameUpdateMsg.Foods, &FoodMsg{
						Id:     food.Id,
						X:      food.X,
						Y:      food.Y,
						Width:  food.Width,
						Height: food.Height,
					})
				}
			}

			for _, other := range g.Snakes {
				if snake.Id != other.Id && !rectCollision(snake.DisplayRect, other.CollisionRect) {
					continue
				}

				gameUpdateMsg.Snakes = append(gameUpdateMsg.Snakes, &SnakeMsg{
					Id:            other.Id,
					X:             other.X,
					Y:             other.Y,
					Width:         other.Width,
					IsSpeedUp:     other.IsSpeedUp,
					FillColor:     other.FillColor,
					Angle:         other.Angle,
					Scale:         other.Scale,
					IsBlinking:    other.IsBlinking,
					CollisionRect: other.CollisionRect,
					Username:      other.Username,
					Speed:         other.Speed,
					Length:        other.Length,
					Score:         int64(len(other.Points)),
					Points:        other.SimplifiedPoints,
				})
			}
		}

		msg := make([]byte, 15000)
		msgLen := gameUpdateMsg.MarshalTo(msg)

		err := conn.WriteMessage(websocket.BinaryMessage, msg[:msgLen])
		if err != nil {
			log.Println("error while sending message", err)
			return
		}
	}
}
